File: frontend/src/api/skills.ts
import { apiClient } from './client';

// Skill type definitions
export interface Skill {
  id: string;
  name: string;
  description: string;
  category: string;
  source: 'official' | 'community' | 'custom';
  icon: string;
  required_tools: string[];
  optional_tools: string[];
  is_premium: boolean;
  default_enabled: boolean;
  version: string;
  author: string | null;
  updated_at: string;
  // Custom skill fields
  owner_id?: string | null;
  is_public?: boolean;
  parent_skill_id?: string | null;
  system_prompt_addition?: string | null;
  // Claude-style configuration fields
  invocation_type?: 'user' | 'ai' | 'both';
  allowed_tools?: string[] | null;
  supports_dynamic_context?: boolean;
  trigger_patterns?: string[];
}

export interface UserSkill {
  skill: Skill;
  enabled: boolean;
  config: Record<string, unknown>;
  order: number;
}

export interface SkillListResponse {
  skills: Skill[];
  total: number;
}

export interface UserSkillsResponse {
  skills: UserSkill[];
  enabled_count: number;
  max_skills: number;
}

export interface UpdateUserSkillRequest {
  enabled?: boolean;
  config?: Record<string, unknown>;
  order?: number;
}

export interface SkillToolsResponse {
  skill_ids: string[];
  tools: string[];
}

/**
 * Get all available skills
 */
export async function getAvailableSkills(category?: string): Promise<Skill[]> {
  const params = category ? { category } : undefined;
  const response = await apiClient.get<{ data: SkillListResponse }>('/skills', { params });
  return response.data.data.skills;
}

/**
 * Get a specific skill by ID
 */
export async function getSkillById(skillId: string): Promise<Skill> {
  const response = await apiClient.get<{ data: Skill }>(`/skills/${skillId}`);
  return response.data.data;
}

/**
 * Get current user's skill configurations
 */
export async function getUserSkills(): Promise<UserSkillsResponse> {
  const response = await apiClient.get<{ data: UserSkillsResponse }>('/skills/user/config');
  return response.data.data;
}

/**
 * Update a user's skill configuration (enable/disable or configure)
 */
export async function updateUserSkill(
  skillId: string,
  data: UpdateUserSkillRequest
): Promise<UserSkill> {
  const response = await apiClient.put<{ data: UserSkill }>(`/skills/user/${skillId}`, data);
  return response.data.data;
}

/**
 * Enable multiple skills at once (replaces current enabled skills)
 */
export async function enableSkills(skillIds: string[]): Promise<UserSkillsResponse> {
  const response = await apiClient.post<{ data: UserSkillsResponse }>(
    '/skills/user/enable',
    { skill_ids: skillIds }
  );
  return response.data.data;
}

/**
 * Get tools required by specific skills
 */
export async function getSkillTools(skillIds: string[]): Promise<SkillToolsResponse> {
  const params = { skill_ids: skillIds.join(',') };
  const response = await apiClient.get<{ data: SkillToolsResponse }>('/skills/tools/required', {
    params,
  });
  return response.data.data;
}

// Skill category enum for frontend use
export const SKILL_CATEGORIES = {
  research: 'Research',
  coding: 'Coding',
  browser: 'Browser',
  file_management: 'File Management',
  data_analysis: 'Data Analysis',
  communication: 'Communication',
  custom: 'Custom',
} as const;

// Maximum skills allowed to be enabled at once
export const MAX_ENABLED_SKILLS = 5;

// Custom Skill CRUD Types
export interface CreateCustomSkillRequest {
  name: string;
  description: string;
  category?: string;
  icon?: string;
  required_tools: string[];
  optional_tools?: string[];
  system_prompt_addition: string;
  // Claude-style configuration (optional)
  invocation_type?: 'user' | 'ai' | 'both';
  allowed_tools?: string[];
  supports_dynamic_context?: boolean;
  trigger_patterns?: string[];
}

export interface UpdateCustomSkillRequest {
  name?: string;
  description?: string;
  icon?: string;
  required_tools?: string[];
  optional_tools?: string[];
  system_prompt_addition?: string;
  // Claude-style configuration updates
  invocation_type?: 'user' | 'ai' | 'both';
  allowed_tools?: string[] | null;
  supports_dynamic_context?: boolean;
  trigger_patterns?: string[];
}

export interface CustomSkillListResponse {
  skills: Skill[];
  total: number;
}

// Custom Skill CRUD Functions

/**
 * Create a new custom skill
 */
export async function createCustomSkill(data: CreateCustomSkillRequest): Promise<Skill> {
  const response = await apiClient.post<{ data: Skill }>('/skills/custom', data);
  return response.data.data;
}

/**
 * Get all custom skills owned by the current user
 */
export async function getMyCustomSkills(): Promise<Skill[]> {
  const response = await apiClient.get<{ data: CustomSkillListResponse }>('/skills/custom');
  return response.data.data.skills;
}

/**
 * Get a specific custom skill by ID
 */
export async function getCustomSkillById(skillId: string): Promise<Skill> {
  const response = await apiClient.get<{ data: Skill }>(`/skills/custom/${skillId}`);
  return response.data.data;
}

/**
 * Update a custom skill
 */
export async function updateCustomSkill(
  skillId: string,
  data: UpdateCustomSkillRequest
): Promise<Skill> {
  const response = await apiClient.put<{ data: Skill }>(`/skills/custom/${skillId}`, data);
  return response.data.data;
}

/**
 * Delete a custom skill
 */
export async function deleteCustomSkill(skillId: string): Promise<void> {
  await apiClient.delete(`/skills/custom/${skillId}`);
}

/**
 * Publish a custom skill to the community
 */
export async function publishCustomSkill(skillId: string): Promise<Skill> {
  const response = await apiClient.post<{ data: Skill }>(`/skills/custom/${skillId}/publish`, {
    confirm: true,
  });
  return response.data.data;
}

// =============================================================================
// SKILL PACKAGE OPERATIONS (for skill delivery)
// =============================================================================

export interface SkillPackageFile {
  path: string;
  content: string;
  size: number;
}

export interface SkillPackageResponse {
  id: string;
  name: string;
  description: string;
  version: string;
  icon: string;
  category: string;
  author?: string;
  file_tree: Record<string, unknown>;
  files: SkillPackageFile[];
  file_id?: string;
  skill_id?: string;
  file_count: number;
  created_at?: string;
}

/**
 * Get a skill package by ID
 */
export async function getSkillPackage(packageId: string): Promise<SkillPackageResponse> {
  const response = await apiClient.get<{ data: SkillPackageResponse }>(
    `/skills/packages/${packageId}`
  );
  return response.data.data;
}

/**
 * Download a skill package as a .skill ZIP file
 */
export async function downloadSkillPackage(packageId: string): Promise<void> {
  const response = await apiClient.get(`/skills/packages/${packageId}/download`, {
    responseType: 'blob',
  });

  // Get filename from Content-Disposition header or use default
  const contentDisposition = response.headers['content-disposition'];
  let filename = 'skill.skill';
  if (contentDisposition) {
    const match = contentDisposition.match(/filename="(.+)"/);
    if (match) {
      filename = match[1];
    }
  }

  // Create blob URL and trigger download
  const blob = response.data as Blob;
  const url = window.URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = url;
  link.download = filename;
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
  window.URL.revokeObjectURL(url);
}

/**
 * Get a single file from a skill package
 */
export async function getSkillPackageFile(
  packageId: string,
  path: string
): Promise<SkillPackageFile> {
  const response = await apiClient.get<{ data: SkillPackageFile }>(
    `/skills/packages/${packageId}/file`,
    { params: { path } }
  );
  return response.data.data;
}

/**
 * Install a skill from a package
 */
export async function installSkillFromPackage(
  packageId: string,
  enableAfterInstall = true
): Promise<Skill> {
  const response = await apiClient.post<{ data: Skill }>(
    `/skills/packages/${packageId}/install`,
    { enable_after_install: enableAfterInstall }
  );
  return response.data.data;
}

File: mockserver/routes/skills.py
from __future__ import annotations

from fastapi import APIRouter, Request
from pydantic import BaseModel
from stores import skill_store
from routes.auth import _get_current_user

router = APIRouter(prefix="/skills")


def _wrap(data):
    return {"code": 0, "msg": "success", "data": data}


@router.get("")
async def list_skills(category: str | None = None):
    skills = skill_store.get_all_skills(category)
    return _wrap({"skills": skills, "total": len(skills)})


@router.get("/user/config")
async def get_user_skills(request: Request):
    user = _get_current_user(request)
    user_skills = skill_store.get_user_skills(user["id"])
    enabled = sum(1 for s in user_skills if s.get("enabled"))
    return _wrap({"skills": user_skills, "enabled_count": enabled, "max_skills": 10})


class UpdateUserSkillRequest(BaseModel):
    enabled: bool | None = None
    config: dict | None = None
    order: int | None = None


@router.put("/user/{skill_id}")
async def update_user_skill(
    skill_id: str, req: UpdateUserSkillRequest, request: Request
):
    user = _get_current_user(request)
    result = skill_store.update_user_skill(
        user["id"], skill_id, req.enabled, req.config, req.order
    )
    if not result:
        return {"code": 404, "msg": "Skill not found", "data": None}
    return _wrap(result)


class EnableSkillsRequest(BaseModel):
    skill_ids: list[str]


@router.post("/user/enable")
async def enable_skills(req: EnableSkillsRequest, request: Request):
    user = _get_current_user(request)
    for sid in req.skill_ids:
        skill_store.update_user_skill(user["id"], sid, enabled=True)
    user_skills = skill_store.get_user_skills(user["id"])
    enabled = sum(1 for s in user_skills if s.get("enabled"))
    return _wrap({"skills": user_skills, "enabled_count": enabled, "max_skills": 10})


@router.get("/tools/required")
async def get_required_tools(skill_ids: str = ""):
    ids = [s.strip() for s in skill_ids.split(",") if s.strip()]
    tools = set()
    for sid in ids:
        skill = skill_store.get_skill(sid)
        if skill:
            tools.update(skill.get("required_tools", []))
    return _wrap({"skill_ids": ids, "tools": list(tools)})


class CreateCustomSkillRequest(BaseModel):
    name: str
    description: str = ""
    category: str = "custom"
    icon: str = "puzzle"
    required_tools: list[str] = []
    optional_tools: list[str] = []
    system_prompt_addition: str = ""
    invocation_type: str = "both"
    allowed_tools: list[str] | None = None
    supports_dynamic_context: bool = False
    trigger_patterns: list[str] = []


@router.post("/custom")
async def create_custom_skill(req: CreateCustomSkillRequest, request: Request):
    user = _get_current_user(request)
    skill = skill_store.create_custom_skill(user["id"], req.model_dump())
    return _wrap(skill)


@router.get("/custom")
async def list_custom_skills(request: Request):
    user = _get_current_user(request)
    skills = skill_store.get_custom_skills(user["id"])
    return _wrap({"skills": skills, "total": len(skills)})


@router.get("/custom/{skill_id}")
async def get_custom_skill(skill_id: str):
    skill = skill_store.get_skill(skill_id)
    if not skill:
        return {"code": 404, "msg": "Skill not found", "data": None}
    return _wrap(skill)


class UpdateCustomSkillRequest(BaseModel):
    name: str | None = None
    description: str | None = None
    icon: str | None = None
    required_tools: list[str] | None = None
    optional_tools: list[str] | None = None
    system_prompt_addition: str | None = None
    invocation_type: str | None = None
    allowed_tools: list[str] | None = None
    supports_dynamic_context: bool | None = None
    trigger_patterns: list[str] | None = None


@router.put("/custom/{skill_id}")
async def update_custom_skill(skill_id: str, req: UpdateCustomSkillRequest):
    skill = skill_store.get_skill(skill_id)
    if not skill:
        return {"code": 404, "msg": "Skill not found", "data": None}
    updates = {k: v for k, v in req.model_dump().items() if v is not None}
    skill.update(updates)
    return _wrap(skill)


@router.delete("/custom/{skill_id}")
async def delete_custom_skill(skill_id: str):
    skill_store.delete_custom_skill(skill_id)
    return _wrap({})


class PublishRequest(BaseModel):
    confirm: bool = True


@router.post("/custom/{skill_id}/publish")
async def publish_custom_skill(skill_id: str, req: PublishRequest):
    skill = skill_store.get_skill(skill_id)
    if not skill:
        return {"code": 404, "msg": "Skill not found", "data": None}
    skill["is_public"] = True
    return _wrap(skill)


@router.get("/packages/{package_id}")
async def get_skill_package(package_id: str):
    return _wrap(
        {
            "id": package_id,
            "name": "Demo Skill Package",
            "description": "A demonstration skill package",
            "version": "1.0.0",
            "icon": "package",
            "category": "demo",
            "author": "Pythinker",
            "file_tree": {"src": {"type": "file", "path": "src/main.py", "size": 256}},
            "files": [
                {
                    "path": "src/main.py",
                    "content": "# Demo skill\nprint('hello')",
                    "size": 256,
                }
            ],
            "file_count": 1,
            "created_at": "2025-12-01T00:00:00Z",
        }
    )


@router.get("/packages/{package_id}/download")
async def download_skill_package(package_id: str):
    from fastapi.responses import Response

    return Response(
        content=b"PK\x03\x04mock_skill_package",
        media_type="application/zip",
        headers={"Content-Disposition": f'attachment; filename="{package_id}.skill"'},
    )


@router.get("/packages/{package_id}/file")
async def get_skill_package_file(package_id: str, path: str = ""):
    return _wrap(
        {
            "path": path,
            "content": "# Demo skill file content\nprint('hello')",
            "size": 128,
        }
    )


class InstallPackageRequest(BaseModel):
    enable_after_install: bool = True


@router.post("/packages/{package_id}/install")
async def install_skill_package(package_id: str, req: InstallPackageRequest):
    # Return a mock installed skill
    return _wrap(
        {
            "id": f"skill_installed_{package_id[:8]}",
            "name": "Installed Skill",
            "description": "Installed from package",
            "category": "custom",
            "source": "custom",
            "icon": "package",
            "required_tools": [],
            "optional_tools": [],
            "is_premium": False,
            "default_enabled": req.enable_after_install,
            "version": "1.0.0",
            "author": None,
            "updated_at": "2025-12-01T00:00:00Z",
        }
    )


# Must be AFTER /custom and /user routes to avoid path conflicts
@router.get("/{skill_id}")
async def get_skill(skill_id: str):
    skill = skill_store.get_skill(skill_id)
    if not skill:
        return {"code": 404, "msg": "Skill not found", "data": None}
    return _wrap(skill)
